sympy/matrices/determinant.py

```diff
@@ -0,0 +1,891 @@
+    return the determinant of the matrix obtained by deleting
+    the `i`th row and `j`th column from ``M``.
+
+    Parameters
+    ==========
+
+    dotprodsimp : bool, optional
+        Specifies whether intermediate term algebraic simplification is used
+        during matrix multiplications to control expression blowup and thus
+        speed up calculation.
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> from sympy.matrices import minor
+    >>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> minor(M, 1, 1)
+    -12
+    >>> minor(M, 1, 1) == M.minor(1, 1)
+    True
+
+    See Also
+    ========
+
+    minor_submatrix
+    cofactor
+    det
+    """
+
+    if not M.is_square:
+        raise NonSquareMatrixError()
+
+    return M.minor_submatrix(i, j).det(method=method, dotprodsimp=dotprodsimp)
+
+
+def _minor_submatrix(M, i, j):
+    """Return the submatrix obtained by removing the `i`th row
+    and `j`th column from ``M`` (works with Pythonic negative indices).
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> from sympy.matrices import minor_submatrix
+    >>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> minor_submatrix(M, 1, 1)
+    Matrix([
+    [1, 3],
+    [7, 9]])
+    >>> minor_submatrix(M, 1, 1) == M.minor_submatrix(1, 1)
+    True
+
+    See Also
+    ========
+
+    minor
+    cofactor
+    """
+
+    if i < 0:
+        i += M.rows
+    if j < 0:
+        j += M.cols
+
+    if not 0 <= i < M.rows or not 0 <= j < M.cols:
+        raise ValueError("`i` and `j` must satisfy 0 <= i < ``M.rows`` "
+                            "(%d)" % M.rows + "and 0 <= j < ``M.cols`` (%d)." % M.cols)
+
+    rows = [a for a in range(M.rows) if a != i]
+    cols = [a for a in range(M.cols) if a != j]
+
+    return M.extract(rows, cols)
+
+
+# The following are top level stand-alone interface functions which sympify the
+# matrix where needed (so it becomes immutable and returns immutable), otherwise
+# the implementations above do not sympify due to problems in other parts of the
+# codebase using matrices of unsympifiable objects. Technically could just
+# assign these via 'charpoly = _charpoly' where sympification is not needed but
+# this is clearer.
+
+def adjugate(M, method="berkowitz", dotprodsimp=None):
+    return sympify(_adjugate(M, method=method, dotprodsimp=dotprodsimp))
+
+def charpoly(M, x='lambda', simplify=_simplify, dotprodsimp=None):
+    return _charpoly(M, x=x, simplify=simplify, dotprodsimp=dotprodsimp)
+
+def cofactor(M, i, j, method="berkowitz", dotprodsimp=None):
+    return _cofactor(M, i, j, method=method, dotprodsimp=dotprodsimp)
+
+def cofactor_matrix(M, method="berkowitz", dotprodsimp=None):
+    return sympify(_cofactor_matrix(M, method=method, dotprodsimp=dotprodsimp))
+
+def det_matrix(M, method="bareiss", iszerofunc=None, dotprodsimp=None):
+    return _det(M, method=method, iszerofunc=iszerofunc, dotprodsimp=dotprodsimp)

```

Comment:
What is the point of having `det_matrix` and `_det`?

Reply:
`_det` is for the matrix class method `.det()` to call. I left `det_matrix` in case the behavior of the top-level functions might be different in the future from class methods (like top level always sympifying to immutable before operation where class methods might not). `_det` also does a quick out for smaller matrix order determinant.
```
        if n == 0:
            return M.one
        elif n == 1:
            return M[0,0]
        elif n == 2:
            m = M[0, 0] * M[1, 1] - M[0, 1] * M[1, 0]
            return _dotprodsimp(m) if dotprodsimp else m
        elif n == 3:
            m =  (M[0, 0] * M[1, 1] * M[2, 2]
                + M[0, 1] * M[1, 2] * M[2, 0]
                + M[0, 2] * M[1, 0] * M[2, 1]
                - M[0, 2] * M[1, 1] * M[2, 0]
                - M[0, 0] * M[1, 2] * M[2, 1]
                - M[0, 1] * M[1, 0] * M[2, 2])
            return _dotprodsimp(m) if dotprodsimp else m
```